### ISSUE TITLE
Semver implications of `#[non_exhaustive]` behavior on tuple/unit enum variants

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#304]
Consider the following enum as the public API of a crate.
```rust
pub enum Foo {
    #[non_exhaustive]
    Unit,

    #[non_exhaustive]
    Tuple(i64),
}
```

Per [the reference](https://doc.rust-lang.org/reference/attributes/type_system.html#the-non_exhaustive-attribute), other crates may not construct nor pattern-match on `Foo::Unit` or `Foo::Tuple`. For struct variants, only constructing them is not allowed; pattern-matching on them is allowed by including `..` in the match pattern.

This appears to have some surprising implications for use of this type from outside its defining crate:
1. Converting a non-exhaustive unit variant to a non-exhaustive tuple variant, or vice versa, is completely invisible i.e. semver-patch. (Optionally, it can be opt-in semver-major if serialization is involved. A good use case for #58.)
2. There's no way to access the fields of the tuple variant. This means that both adding *and removing* fields is non-breaking — in fact it's semver-patch because it's completely invisible. (Same caveat about serialization as above.)
3. As written, there's no reasonable way to even determine which of the two variants is inside. Sure, `std::mem::discriminant()` can return a value, but in isolation that value means nothing — [its ordering is specified but the exact value is not](https://rust-lang.github.io/rfcs/0639-discriminant-intrinsic.html). As written, *deleting* a non-exhaustive unit or tuple variant is invisible and therefore semver-patch.

The only way I've found to externally distinguish between `Foo::Unit` and `Foo::Tuple` requires explicit discriminants so that `std::mem::discriminant()`'s return value can be defined exactly rather than just in terms of its ordering:
```rust
#[repr(C)]  // or any explicit int repr
pub enum Foo {
    #[non_exhaustive]
    Unit = 1,

    #[non_exhaustive]
    Tuple(i64) = 2,
}
```

Another question: is reordering enum variants considered semver-major, minor, or patch? It definitely could change the outcomes of use of  `std::mem::discriminant()`, but some rare technically-might-be-breaking changes are not considered semver-major (e.g. adding a method) so I feel this isn't a clear-cut case. The [API evolution RFC](https://rust-lang.github.io/rfcs/1105-api-evolution.html) does not address this either.

**Comment by obi1kenobi** [comment]
The following syntax is able to match on the variants and access the tuple variant's fields:
```rust
match foo {
    Foo::Unit { .. } => {},
    Foo::Tuple { 0: val, .. } => {},
}
```
h/t to https://old.reddit.com/r/rust/comments/10h9my4/semver_implications_of_non_exhaustive_behavior_on/j57d5cf/

**Comment by Aloso** [comment]
You can also use `..` in a tuple variant pattern:

```rust
match foo {
    Foo::Tuple { 0: val, .. } => {},
    // is equivalent to
    Foo::Tuple(val, ..) => {},
}
```

**Comment by obi1kenobi** [comment]
That's what I thought at first too, but it didn't seem to work: on rustc 1.68 (nightly) I got an `error[E0603]: tuple variant `Tuple` is private` error.

**Comment by obi1kenobi** [comment]
Opened a diagnostic issue on rustc: https://github.com/rust-lang/rust/issues/107165
